### PR TITLE
RM#12488: Can now generate project with task templates by product

### PR DESCRIPTION
**Pull request by fsc-axelor**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -22,6 +23,7 @@
 - Timeshee editor: Allows to delete a line when use unique product for the timesheet. 
 - Project Folder: The name is now display like title. Disable Project edit and new from Folder. 
 - Automatic mail default fields in stock move are now filled correctly on partner change.
+- Project: Fix multiple bugs and increase views.
```
Comment (ale-axelor) [inline CHANGELOG.md]: "increase views" → "improve views".

File: axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/ProductTaskTemplateServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      ProductTaskTemplateRepository repository,
+      TeamTaskBusinessService teamTaskBusinessService,
+      TeamTaskRepository teamTaskRepository) {
+    this.repository = repository;
+    this.teamTaskBusinessService = teamTaskBusinessService;
+    this.teamTaskRepository = teamTaskRepository;
+  }
+
+  @Override
+  @Transactional(rollbackOn = {AxelorException.class, Exception.class})
```
Comment (ale-axelor) [inline axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/ProductTaskTemplateServiceImpl.java]: It works, but it makes more sense here to write `@Transactional`, as no axelor exception is thrown. See https://github.com/google/guice/wiki/Transactions#responding-to-exceptions

File: axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/ProductTaskTemplateServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+        qtyTmp = qtyTmp.subtract(BigDecimal.ONE);
+      }
+    }
+
+    return tasks;
+  }
+
+  @Override
+  @Transactional(rollbackOn = {AxelorException.class, Exception.class})
```
Comment (ale-axelor) [inline axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/ProductTaskTemplateServiceImpl.java]: Same as above, it's better to write `@Transactional` here.

File: axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/ProjectGeneratorFactory.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      case TASK_BY_LINE:
+        return Beans.get(ProjectGeneratorFactoryTask.class);
+      case PHASE_BY_LINE:
+        return Beans.get(ProjectGeneratorFactoryPhase.class);
+      case TASK_TEMPLATE:
+        return Beans.get(ProjectGeneratorFactoryTaskTemplate.class);
+      default:
+        throw new AxelorException(
+            TraceBackRepository.TYPE_FUNCTIONNAL,
+            I18n.get("Factory not found this type of generator"));
```
Comment (ale-axelor) [inline axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/ProjectGeneratorFactory.java]: It's better to put exception text into seperated variable for reusability.
"Factory not found this type of generator" → "Factory not found for this type of generator"

File: axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/factory/ProjectGeneratorFactoryAlone.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+  @Inject
+  public ProjectGeneratorFactoryAlone(
+      ProjectBusinessService projectBusinessService, ProjectRepository projectRepository) {
+    this.projectBusinessService = projectBusinessService;
+    this.projectRepository = projectRepository;
+  }
+
+  @Override
+  @Transactional(rollbackOn = {AxelorException.class, Exception.class})
```
Comment (ale-axelor) [inline axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/factory/ProjectGeneratorFactoryAlone.java]: Same as above, it's better to write `@Transactional` here.

File: axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/factory/ProjectGeneratorFactoryAlone.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    project.setIsBusinessProject(true);
+    return projectRepository.save(project);
+  }
+
+  @Override
+  public ActionViewBuilder fill(Project project, SaleOrder saleOrder, LocalDateTime localDateTime)
+      throws AxelorException {
+    throw new AxelorException(
+        TraceBackRepository.TYPE_FUNCTIONNAL,
+        I18n.get("You can't fill a project with the strategy Project Alone."));
```
Comment (ale-axelor) [inline axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/factory/ProjectGeneratorFactoryAlone.java]: It's better to put exception text into seperated variable for reusability.

File: axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/factory/ProjectGeneratorFactoryPhase.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+      ProjectBusinessService projectBusinessService,
+      ProjectRepository projectRepository,
+      SaleOrderLineRepository saleOrderLineRepository) {
+    this.projectBusinessService = projectBusinessService;
+    this.projectRepository = projectRepository;
+    this.saleOrderLineRepository = saleOrderLineRepository;
+  }
+
+  @Override
+  @Transactional(rollbackOn = {AxelorException.class, Exception.class})
```
Comment (ale-axelor) [inline axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/factory/ProjectGeneratorFactoryPhase.java]: Same as above, it's better to write `@Transactional` here.

File: axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/factory/ProjectGeneratorFactoryPhase.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  public Project create(SaleOrder saleOrder) {
+    Project project = projectBusinessService.generateProject(saleOrder);
+    project.setIsProject(false);
+    project.setIsBusinessProject(true);
+    return projectRepository.save(project);
+  }
+
+  @Override
+  @Transactional(rollbackOn = {AxelorException.class, Exception.class})
```
Comment (ale-axelor) [inline axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/factory/ProjectGeneratorFactoryPhase.java]: Same as above, it's better to write `@Transactional` here.

File: axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/factory/ProjectGeneratorFactoryTask.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      TeamTaskBusinessService teamTaskBusinessService,
+      TeamTaskRepository teamTaskRepository) {
+    this.projectBusinessService = projectBusinessService;
+    this.projectRepository = projectRepository;
+    this.teamTaskBusinessService = teamTaskBusinessService;
+    this.teamTaskRepository = teamTaskRepository;
+  }
+
+  @Override
+  @Transactional(rollbackOn = {AxelorException.class, Exception.class})
```
Comment (ale-axelor) [inline axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/factory/ProjectGeneratorFactoryTask.java]: Same as above, it's better to write `@Transactional` here.

File: axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/factory/ProjectGeneratorFactoryTask.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  public Project create(SaleOrder saleOrder) {
+    Project project = projectBusinessService.generateProject(saleOrder);
+    project.setIsProject(true);
+    project.setIsBusinessProject(true);
+    return projectRepository.save(project);
+  }
+
+  @Override
+  @Transactional(rollbackOn = {AxelorException.class, Exception.class})
```
Comment (ale-axelor) [inline axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/factory/ProjectGeneratorFactoryTask.java]: Same as above, it's better to write `@Transactional` here.

File: axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/factory/ProjectGeneratorFactoryTaskTemplate.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+      ProductTaskTemplateService productTaskTemplateService) {
+    this.projectBusinessService = projectBusinessService;
+    this.projectRepository = projectRepository;
+    this.teamTaskService = teamTaskService;
+    this.teamTaskRepository = teamTaskRepository;
+    this.productTaskTemplateService = productTaskTemplateService;
+  }
+
+  @Override
+  @Transactional(rollbackOn = {AxelorException.class, Exception.class})
```
Comment (ale-axelor) [inline axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/factory/ProjectGeneratorFactoryTaskTemplate.java]: Same as above, it's better to write `@Transactional` here.

File: axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/factory/ProjectGeneratorFactoryTaskTemplate.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+  public Project create(SaleOrder saleOrder) {
+    Project project = projectBusinessService.generateProject(saleOrder);
+    project.setIsProject(true);
+    project.setIsBusinessProject(true);
+    return projectRepository.save(project);
+  }
+
+  @Override
+  @Transactional(rollbackOn = {AxelorException.class, Exception.class})
```
Comment (ale-axelor) [inline axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/projectgenerator/factory/ProjectGeneratorFactoryTaskTemplate.java]: Same as above, it's better to write `@Transactional` here.

File: axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/ProductTaskTemplateServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+      Project project,
+      TeamTask parent,
+      LocalDateTime startDate,
+      BigDecimal qty) {
+    List<TeamTask> tasks = new ArrayList<>();
+
+    for (TaskTemplate template : templates) {
+      BigDecimal qtyTmp = (template.getIsUniqueTaskForMultipleQuantity() ? BigDecimal.ONE : qty);
+
+      while (qtyTmp.signum() != 0) {
```
Comment (ale-axelor) [inline axelor-business-project/src/main/java/com/axelor/apps/businessproject/service/ProductTaskTemplateServiceImpl.java]: qtyTmp.signum() > 0 or else there an infinite loop could happen.

File: axelor-project/src/main/resources/domains/TaskTemplate.xml [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+  <entity name="TaskTemplate">
+    <string name="name" namecolumn="true" title="Name" search="id,name"/>
+    <decimal name="delayToStart" default="0" title="Delay to start (Hours)"/>
+    <decimal name="duration" title="Duration (Hours)"/>
+    <decimal name="totalPlannedHrs" title="Total planned hours"/>
+    <many-to-one name="parentTaskTemplate" ref="com.axelor.apps.project.db.TaskTemplate"
+      title="Parent task"/>
+    <one-to-many name="taskTemplateList" ref="com.axelor.apps.project.db.TaskTemplate"
+      mappedBy="parentTaskTemplate" orphanRemoval="true"/>
+    <many-to-one name="assignedTo" ref="com.axelor.auth.db.User" title="Assigned to"/>
```
Comment (gdu-axelor) [inline axelor-project/src/main/resources/domains/TaskTemplate.xml]: assignedToUser

**Pull request merged**